### PR TITLE
fix: add expected errors in MariaDB

### PR DESCRIPTION
**Pull request by ZhengLin-Li**
fix: add expected errors in MariaDB

### CODE REVIEW

**Comment by bajinsheng** [comment]
It seems SQLancer triggers another issue in MariaDB?

**Comment by ZhengLin-Li** [comment]
I reproduced the queries on Ubuntu 22.04 and found the following:
image

It seams that table `t0` does not support optimize. How do we deal with this?

**Comment by bajinsheng** [comment]
1: I think it can be solved by checking whether the table uses the InnoDB storage engine. If it is difficult, we can simply ignore it as it is not a significant issue.
2. Perhaps you can try to run the full log, like database0-cur.log? I guess some statements that are ignored in the crash log, may incur this issue.

**Comment by ZhengLin-Li** [comment]
> 1: I think it can be solved by checking whether the table uses the InnoDB storage engine. If it is difficult, we can simply ignore it as it is not a significant issue. 2. Perhaps you can try to run the full log, like database0-cur.log? I guess some statements that are ignored in the crash log, may incur this issue.

1. Yes, it uses InnoDB engine. But how which engine will affact optimize statement?
2. Below is the database11.log
```sql
--java.lang.AssertionError: INSERT INTO t0 VALUES (-1497191741);
--	at sqlancer.common.query.SQLQueryAdapter.checkException(SQLQueryAdapter.java:111)
--	at sqlancer.common.query.SQLQueryAdapter.execute(SQLQueryAdapter.java:93)
--	at sqlancer.Main$QueryManager.execute(Main.java:265)
--	at sqlancer.GlobalState.executeStatement(GlobalState.java:108)
--	at sqlancer.mariadb.MariaDBProvider.generateDatabase(MariaDBProvider.java:151)
--	at sqlancer.mariadb.MariaDBProvider.generateDatabase(MariaDBProvider.java:1)
--	at sqlancer.ProviderAdapter.generateAndTestDatabase(ProviderAdapter.java:52)
--	at sqlancer.Main$DBMSExecutor.run(Main.java:364)
--	at sqlancer.Main$2.run(Main.java:559)
--	at sqlancer.Main$2.runThread(Main.java:541)
--	at sqlancer.Main$2.run(Main.java:532)
--	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
--	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
--	at java.base/java.lang.Thread.run(Thread.java:829)
--Caused by: java.sql.SQLSyntaxErrorException: (conn=570) Table 'database11.t0' doesn't exist
--	at org.mariadb.jdbc.export.ExceptionFactory.createException(ExceptionFactory.java:282)
--	at org.mariadb.jdbc.export.ExceptionFactory.create(ExceptionFactory.java:370)
--	at org.mariadb.jdbc.message.ClientMessage.readPacket(ClientMessage.java:134)
--	at org.mariadb.jdbc.client.impl.StandardClient.readPacket(StandardClient.java:855)
--	at org.mariadb.jdbc.client.impl.StandardClient.readResults(StandardClient.java:794)
--	at org.mariadb.jdbc.client.impl.StandardClient.readResponse(StandardClient.java:713)
--	at org.mariadb.jdbc.client.impl.StandardClient.execute(StandardClient.java:637)
--	at org.mariadb.jdbc.Statement.executeInternal(Statement.java:941)
--	at org.mariadb.jdbc.Statement.execute(Statement.java:1067)
--	at org.mariadb.jdbc.Statement.execute(Statement.java:458)
--	at sqlancer.common.query.SQLQueryAdapter.execute(SQLQueryAdapter.java:87)
--	... 12 more
---- Time: 2023/06/21 12:30:12
-- Database: database11
-- Database version: 10.3.38-MariaDB-0ubuntu0.20.04.1
-- seed value: 11
DROP DATABASE IF EXISTS database11;
CREATE DATABASE database11;
USE database11;
CREATE OR REPLACE TABLE t0(c0 BOOLEAN  UNIQUE);
SET SESSION myisam_stats_method = nulls_unequal;
SET SESSION join_cache_level = 5;
INSERT INTO t0 VALUES (299405238);
CHECKSUM TABLE t0;
INSERT INTO t0 VALUES (-486923179);
INSERT INTO t0 VALUES (-471262417);
SET SESSION optimizer_search_depth = 4;
INSERT INTO t0 VALUES (618607252);
INSERT INTO t0 VALUES (-433413318);
INSERT INTO t0 VALUES (266645617);
SET SESSION max_length_for_sort_data = 3932438;
SET SESSION sql_auto_is_null = ON;
INSERT INTO t0 VALUES (916765166);
SET SESSION sql_log_off = ON;
SET SESSION max_sort_length = 6112935;
INSERT INTO t0 VALUES (550488471);
SET SESSION optimizer_prune_level = 0;
INSERT INTO t0 VALUES (499498418);
INSERT INTO t0 VALUES (1540761514);
SET SESSION autocommit = 1;
SET SESSION unique_checks = ON;
SET SESSION join_cache_level = 5;
INSERT INTO t0 VALUES (1476203969);
INSERT INTO t0 VALUES (-2079398998);
SET SESSION join_cache_level = 1;
INSERT INTO t0 VALUES (-839316519);
TRUNCATE t0  WAIT 1;
SET SESSION sql_buffer_result = OFF;
INSERT INTO t0 VALUES (1712230220);
SET SESSION sql_quote_show_create = OFF;
INSERT INTO t0 VALUES (1984650463);
INSERT INTO t0 VALUES (NULL);

```


In table `t0`, we have only one boolean column, but we insert a value of Integer? It does not meet expectation.

**Comment by bajinsheng** [comment]
1 I guess only innodb engines can not use optimization, so perhaps you can use storage engine to identify it.
2 Is it the full log? Sometimes the column data types are weak, and accept different data types. Or the staement is rejected by database and sqlancer does not show the error information.

**Comment by ZhengLin-Li** [comment]
It seams that the error message in CI is different from before, now the only CI error message is:
`AssertionError: Lock wait timeout exceeded; try restarting transaction`

**Comment by ZhengLin-Li** [comment]
I tested it on ubuntu 22.04 and it result in build success.
image

I will test for serveral times and see what's the result.

**Comment by ZhengLin-Li** [comment]
@bajinsheng could you return this workflow accroding to https://docs.github.com/en/actions/managing-workflow-runs/re-running-workflows-and-jobs. I can not see the re-run bottom.

**Comment by bajinsheng** [comment]
Done.

**Comment by ZhengLin-Li** [comment]
I can not reproduce the error besides GitHub Action.

Acutally I found there are too many errors in the SQL queries we excuted. For example, we intern a `null` value to a `NOT NULL` column. And we optimize a InnoDB engine table. 

I think the optimize operation is the cause but can not determine the root cause. How about adding one `SHOW ENGINE INNODB STATUS` statement before the optimize operation, this will let us see the InnoDB engine status? Accroding to https://stackoverflow.com/a/5837308

**Comment by mrigger** [comment]
> Acutally I found there are too many errors in the SQL queries we excuted. For example, we intern a null value to a NOT NULL column. And we optimize a InnoDB engine table.

I think we are making a deliberate tradeoff here. Checking for all these different things is quite tedious, and by inserting invalid values, we can also test the DBMSs. Also see https://github.com/sqlancer/sqlancer/blob/main/CONTRIBUTING.md#expected-errors regarding data types.

**Comment by mrigger** [comment]
> I tested it on ubuntu 22.04 and it result in build success FOR TWO TIMES.

Did you use the same MariaDB version we use in the CI? I assume you also copied the statements for running the tests?

**Comment by ZhengLin-Li** [comment]
> > I tested it on ubuntu 22.04 and it result in build success FOR TWO TIMES.
> 
> Did you use the same MariaDB version we use in the CI? I assume you also copied the statements for running the tests?

Yes, the same version. Yes, copied the logs from CI.

I tested again and found the reason for these two seccuss https://github.com/sqlancer/sqlancer/pull/834#issuecomment-1602677196 here is that, SQLancer on my local mechine ran 0 queries and 0 was succesful, resulting the build success.

**Comment by mrigger** [comment]
>SQLancer on my local mechine ran 0 queries and 0 was succesful, resulting the build success.

It seems like some kind of deadlock is triggered?

**Comment by ZhengLin-Li** [comment]
I tested again in GitHub's Codespace (Ubuntu 22.04) and got 0 queries executed again. I am wondering if there are any errors with my commands. Could anyone help take a look?

The commands I run are:

```bash
sudo apt-key adv --recv-keys --keyserver hkp://keyserver.ubuntu.com:80 0xF1656F24C74CD1D8
sudo add-apt-repository 'deb [arch=amd64,arm64,ppc64el] http://sfo1.mirrors.digitalocean.com/mariadb/repo/10.3/ubuntu bionic main'   
sudo apt update
sudo apt install mariadb-server
sudo service mysql start
sudo mysql -uroot -proot -e "CREATE USER 'sqlancer'@'localhost' IDENTIFIED BY 'sqlancer'; GRANT ALL PRIVILEGES ON * . * TO 'sqlancer'@'localhost';"

git clone https://github.com/sqlancer/sqlancer.git
cd sqlancer
MARIADB_AVAILABLE=true mvn -Dtest=TestMariaDB test
```

**Comment by ZhengLin-Li** [comment]
> > SQLancer on my local mechine ran 0 queries and 0 was succesful, resulting the build success.
> 
> It seems like some kind of deadlock is triggered?

Not sure what's the reason. Could you help take a look at the commands I ran above?

**Comment by bajinsheng** [comment]
It looks like a deadlock bug. If so, it is hard to bypass this bug in the CI.

**Comment by mrigger** [comment]
According to https://github.com/sqlancer/sqlancer/blob/1f80de16e38c4a144e8dabbcef00d07cd15470f9/test/sqlancer/dbms/TestMariaDB.java#L18, it seems expected that we don't generate any queries. I remember that queries often crashed MariaDB, so this might be the reason why we disabled it. So perhaps this is an issue with multiple threads and generating databases? @ZhengLin-Li, perhaps you could try to set the number of threads to 1 here, to see if we can avoid the deadlock issue?

**Comment by ZhengLin-Li** [comment]
> It looks like a deadlock bug. If so, it is hard to bypass this bug in the CI.

You mean the 0 query bug? I solved it by running setting `--num-threads 1`. This bug only happens on my Ubuntu, and it did not occur in CI.




> According to
> 
> https://github.com/sqlancer/sqlancer/blob/1f80de16e38c4a144e8dabbcef00d07cd15470f9/test/sqlancer/dbms/TestMariaDB.java#L18
> 
> , it seems expected that we don't generate any queries. I remember that queries often crashed MariaDB, so this might be the reason why we disabled it. So perhaps this is an issue with multiple threads and generating databases? @ZhengLin-Li, perhaps you could try to set the number of threads to 1 here, to see if we can avoid the deadlock issue?

Oh, yes, setting `--num-threads 1` solved the issue. I tested it again with the command `java -jar sqlancer-*.jar --num-threads 1 mariadb` and it can successfully run queries. 

But the problem is that I still cannot reproduce the `Lock wait timeout exceeded` errors as those in CI.

I analyzed the CI log and found all assertion errors were triggered by the optimize table statement, but all optimize table statements on my Ubuntu work fine.

**Comment by mrigger** [comment]
In this case, what about disabling `OPTIMIZE` for now using the pattern described in https://github.com/sqlancer/sqlancer/blob/main/CONTRIBUTING.md#unfixed-bugs?

**Comment by ZhengLin-Li** [comment]
> In this case, what about disabling `OPTIMIZE` for now using the pattern described in https://github.com/sqlancer/sqlancer/blob/main/CONTRIBUTING.md#unfixed-bugs?

So in this case, we need to report this issue to the MariaDB team and refer to their issue page?

**Comment by bajinsheng** [comment]
If you can produce a minimal test case to reproduce the issue, say executing several SQL statements triggers this deadlock, you can report it to MySQL.

**Comment by mrigger** [comment]
> > In this case, what about disabling `OPTIMIZE` for now using the pattern described in https://github.com/sqlancer/sqlancer/blob/main/CONTRIBUTING.md#unfixed-bugs?
> 
> So in this case, we need to report this issue to the MariaDB team and refer to their issue page?

This would be ideal, as suggested by @bajinsheng, but we could also just refer to this PR for now (or create a separate issue in the repo so someone can investigate it in the future).

**Comment by ZhengLin-Li** [comment]
Do not know why it is still reporting errors after we using  https://github.com/sqlancer/sqlancer/blob/main/CONTRIBUTING.md#unfixed-bugs

**Comment by bajinsheng** [comment]
This error happens in the `TRUNCATE` statement, so you should add expected errors to https://github.com/sqlancer/sqlancer/blob/main/src/sqlancer/mariadb/gen/MariaDBTruncateGenerator.java
Otherwise, the error can not be caught/

File: src/sqlancer/mariadb/gen/MariaDBIndexGenerator.java [hunk 1]
```diff
@@ -19,6 +20,12 @@ public static SQLQueryAdapter generate(MariaDBSchema s) {
         ExpectedErrors errors = new ExpectedErrors();
         StringBuilder sb = new StringBuilder("CREATE ");
         errors.add("Key/Index cannot be defined on a virtual generated column");
+
+        if (MariaDBBugs.bug835) {
+            errors.add("Specified key was too long; max key length is 2300 bytes");
+            errors.add("Lock wait timeout exceeded; try restarting transaction");
```
[the remark this answers is not in the document]
Comment (ZhengLin-Li) [reply src/sqlancer/mariadb/gen/MariaDBIndexGenerator.java:26]: 1. avoiding the generation of such statements also make sense. will try to do this
2. `Specified key was too long; max key length is 2300 bytes` is a bug here: https://github.com/sqlancer/sqlancer/actions/runs/5357089828/jobs/9717469891

image

[the remark this answers is not in the document]
Comment (mrigger) [reply src/sqlancer/mariadb/gen/MariaDBIndexGenerator.java:26]: I think this looks like a new limitation? In this case, we shouldn't consider it as a bug, but just add it as an expected error.

**Comment by bajinsheng** [comment]
You disabled the `TRUNCATE` statements? I meant adding expected errors to the generator of `TRUNCATE1

**Comment by mrigger** [comment]
> You disabled the `TRUNCATE` statements? I meant adding expected errors to the generator of `TRUNCATE1

Does this statement result in a deadlock? If so, we might want to avoid generating the statement, since otherwise, SQLancer cannot generate any additional tests. If not, adding an expected error might be better.

File: src/sqlancer/mariadb/MariaDBProvider.java [hunk 2]
```diff
@@ -126,6 +129,9 @@ public void generateDatabase(MariaDBGlobalState globalState) throws Exception {
                     query = MariaDBInsertGenerator.insert(globalState.getSchema(), globalState.getRandomly());
                     break;
                 case OPTIMIZE:
+                    if (MariaDBBugs.bug834) {
+                        break;
```
Comment (mrigger) [inline src/sqlancer/mariadb/MariaDBProvider.java:133]: The logic here looks and below quite complex. Can we just set `nrPerformed` to `0`?

**Comment by ZhengLin-Li** [comment]
> > You disabled the `TRUNCATE` statements? I meant adding expected errors to the generator of `TRUNCATE1
> 
> Does this statement result in a deadlock? If so, we might want to avoid generating the statement, since otherwise, SQLancer cannot generate any additional tests. If not, adding an expected error might be better.

After analsis, I think such a `TRUNCATE` statement will not result in deadlock, so maybe we can add it to the expected errors.

**Comment by ZhengLin-Li** [comment]
It turned out that, presently, we got:

- 198 `java.lang.AssertionError: `
    - 46 `java.lang.AssertionError: Got error -2 "Internal error < 0 (Not system error)" from storage engine InnoDB`
    - 148 `java.lang.AssertionError: Operation failed`
    - 4 `java.lang.AssertionError: Error updating stats for table 't0' after table rebuild: Lock wait timeout`

**Comment by mrigger** [comment]
Some of these seem like potential bugs that we should look into and potentially report. I guess this might take a while, so not sure if you might want to focus on StoneDB first?

**Comment by bajinsheng** [comment]
I asked @ZhengLin-Li to fix this ERROR in CI so that I can also merge his PR. Since this bug is hard to fix and you can help to merge pr, I think we can lower its priority now.

**Comment by ZhengLin-Li** [comment]
ok, I will just add it to the todo list.

**Comment by malwaregarry** [comment]
thanks @ZhengLin-Li. The bugs were added in this PR https://github.com/sqlancer/sqlancer/pull/945

**Pull request closed without merging**